numpy/_core/src/umath/stringdtype_ufuncs.cpp

```diff
@@ -1028,6 +1028,25 @@ all_strings_promoter(PyObject *NPY_UNUSED(ufunc),
                      PyArray_DTypeMeta *const signature[],
                      PyArray_DTypeMeta *new_op_dtypes[])
 {
+    if (op_dtypes[0] != &PyArray_StringDType &&
+            op_dtypes[1] != &PyArray_StringDType &&
+            op_dtypes[2] != &PyArray_StringDType) {
+        /*
+         * This promoter was triggered with only unicode arguments, so use
```

Comment:
This seems confusing - should `StringDType` really decide the result for something that does not include itself? Shouldn't that be up to `UnicodeDType`? What happens if we return -1 here?

Reply:
Things would just fail the operation, I am very sure I added it for a reason.  The problem is that you would have to decide that `UU->U` is clearly better than `UU->T`.  But there is nothing to decide that by, so the machinery prefers the promoter, since the promoter has the ability to resolve which one to actual use (also go to `UU->U`).

The other thing you might not like is that it matches at all, but that would need one of two new features (which is fine):
* Always call promoters right away even if there may be much better matches and give them the ability to say "don't know" (i.e. a -1 return might be that without an error set).
* Allow some heuristic for it like, "must contain this DType" as an additional dtype.
* Allow "matching" via a second function.

The other solution for the particular case is that if there wasn't legacy promotion involved, I would like a default promoter that ensures that `ufunc(..., dtype=X)` will search for `ufunc(..., signature=(X, X, X))` as well if there is otherwise no match.  That would do the right thing here.

Reply:
It was probably good to get this in, but it still feels weird to have a promotor for a given dtype return a result that does not involve that type at all - how can it decide for another type what is acceptable? Two of your solutions sound reasonable: returning the equivalent of `NotImplemented` (your `-1` with no error set), or a default promotor. Maybe worth a new issue?